[Clang] warning: use of the 'nodiscard' attribute is a C++17 extension [-Wc++17-extensions]

- What is the issue you have?

When compiling on Clang (tested 3.9.1 - 7.0.0) with the following compiler flags:
`-Wpedantic -std=c++11`

Clang displays warnings about:
`warning: use of the 'nodiscard' attribute is a C++17 extension [-Wc++17-extensions]`

Full log:
```
json.hpp:13087:5: error: use of the 'nodiscard' attribute is a C++17 extension [-Werror,-Wc++17-extensions]
    JSON_NODISCARD
    ^
json.hpp:496:34: note: expanded from macro 'JSON_NODISCARD'
        #define JSON_NODISCARD [[nodiscard]]
                                 ^
json.hpp:14262:5: error: use of the 'nodiscard' attribute is a C++17 extension [-Werror,-Wc++17-extensions]
    JSON_NODISCARD
    ^
json.hpp:496:34: note: expanded from macro 'JSON_NODISCARD'
        #define JSON_NODISCARD [[nodiscard]]
                                 ^
json.hpp:14306:5: error: use of the 'nodiscard' attribute is a C++17 extension [-Werror,-Wc++17-extensions]
    JSON_NODISCARD
    ^
json.hpp:496:34: note: expanded from macro 'JSON_NODISCARD'
        #define JSON_NODISCARD [[nodiscard]]
                                 ^
json.hpp:18858:5: error: use of the 'nodiscard' attribute is a C++17 extension [-Werror,-Wc++17-extensions]
    JSON_NODISCARD
    ^
json.hpp:496:34: note: expanded from macro 'JSON_NODISCARD'
        #define JSON_NODISCARD [[nodiscard]]
                                 ^
json.hpp:19627:5: error: use of the 'nodiscard' attribute is a C++17 extension [-Werror,-Wc++17-extensions]
    JSON_NODISCARD
    ^
json.hpp:496:34: note: expanded from macro 'JSON_NODISCARD'
        #define JSON_NODISCARD [[nodiscard]]
                                 ^
json.hpp:19643:5: error: use of the 'nodiscard' attribute is a C++17 extension [-Werror,-Wc++17-extensions]
    JSON_NODISCARD
    ^
json.hpp:496:34: note: expanded from macro 'JSON_NODISCARD'
        #define JSON_NODISCARD [[nodiscard]]
                                 ^
json.hpp:19736:5: error: use of the 'nodiscard' attribute is a C++17 extension [-Werror,-Wc++17-extensions]
    JSON_NODISCARD
    ^
json.hpp:496:34: note: expanded from macro 'JSON_NODISCARD'
        #define JSON_NODISCARD [[nodiscard]]
                                 ^
json.hpp:19752:5: error: use of the 'nodiscard' attribute is a C++17 extension [-Werror,-Wc++17-extensions]
    JSON_NODISCARD
    ^
json.hpp:496:34: note: expanded from macro 'JSON_NODISCARD'
        #define JSON_NODISCARD [[nodiscard]]
                                 ^
json.hpp:19824:5: error: use of the 'nodiscard' attribute is a C++17 extension [-Werror,-Wc++17-extensions]
    JSON_NODISCARD
    ^
json.hpp:496:34: note: expanded from macro 'JSON_NODISCARD'
        #define JSON_NODISCARD [[nodiscard]]
                                 ^
json.hpp:19840:5: error: use of the 'nodiscard' attribute is a C++17 extension [-Werror,-Wc++17-extensions]
    JSON_NODISCARD
    ^
json.hpp:496:34: note: expanded from macro 'JSON_NODISCARD'
        #define JSON_NODISCARD [[nodiscard]]
                                 ^
json.hpp:19911:5: error: use of the 'nodiscard' attribute is a C++17 extension [-Werror,-Wc++17-extensions]
    JSON_NODISCARD
    ^
json.hpp:496:34: note: expanded from macro 'JSON_NODISCARD'
        #define JSON_NODISCARD [[nodiscard]]
                                 ^
json.hpp:19927:5: error: use of the 'nodiscard' attribute is a C++17 extension [-Werror,-Wc++17-extensions]
    JSON_NODISCARD
    ^
json.hpp:496:34: note: expanded from macro 'JSON_NODISCARD'
        #define JSON_NODISCARD [[nodiscard]]
                                 ^
json.hpp:20516:5: error: use of the 'nodiscard' attribute is a C++17 extension [-Werror,-Wc++17-extensions]
    JSON_NODISCARD
    ^
json.hpp:496:34: note: expanded from macro 'JSON_NODISCARD'
        #define JSON_NODISCARD [[nodiscard]]
                                 ^
13 errors generated.
```

- Please describe the steps to reproduce the issue. Can you provide a small but working code example?

Compile json.hpp (3.6.1) using Clang (tested 3.9.1 - 7.0.0) with the following compiler flags:
`-Wpedantic -std=c++11`

- What is the expected behavior?

Compilation succeeds without error / warning?

- And what is the actual behavior instead?

See above.

- Which compiler and operating system are you using? Is it a [supported compiler](https://github.com/nlohmann/json#supported-compilers)?

Clang (tested 3.9.1 - 7.0.0)

- Did you use a released version of the library or the version from the `develop` branch?

Release 3.6.1

- If you experience a compilation error: can you [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests)?

--- a/include/nlohmann/detail/macro_scope.hpp
+++ b/include/nlohmann/detail/macro_scope.hpp
@@ -18,6 +18,14 @@
     #endif
 #endif
 
+// C++ language standard detection
+#if (defined(__cplusplus) && __cplusplus >= 201703L) || (defined(_HAS_CXX17) && _HAS_CXX17 == 1) // fix for issue #464
+    #define JSON_HAS_CPP_17
+    #define JSON_HAS_CPP_14
+#elif (defined(__cplusplus) && __cplusplus >= 201402L) || (defined(_HAS_CXX14) && _HAS_CXX14 == 1)
+    #define JSON_HAS_CPP_14
+#endif
+
 // disable float-equal warnings on GCC/clang
 #if defined(__clang__) || defined(__GNUC__) || defined(__GNUG__)
     #pragma GCC diagnostic push
@@ -42,7 +50,11 @@
 // allow for portable nodiscard warnings
 #if defined(__has_cpp_attribute)
     #if __has_cpp_attribute(nodiscard)
-        #define JSON_NODISCARD [[nodiscard]]
+        #if defined(__clang__) && !defined(JSON_HAS_CPP_17) // issue #1535
+            #define JSON_NODISCARD
+        #else
+            #define JSON_NODISCARD [[nodiscard]]
+        #endif
     #elif __has_cpp_attribute(gnu::warn_unused_result)
         #define JSON_NODISCARD [[gnu::warn_unused_result]]
     #else
@@ -93,14 +105,6 @@
 #else
     #define JSON_LIKELY(x)      x
     #define JSON_UNLIKELY(x)    x
-#endif
-
-// C++ language standard detection
-#if (defined(__cplusplus) && __cplusplus >= 201703L) || (defined(_HAS_CXX17) && _HAS_CXX17 == 1) // fix for issue #464
-    #define JSON_HAS_CPP_17
-    #define JSON_HAS_CPP_14
-#elif (defined(__cplusplus) && __cplusplus >= 201402L) || (defined(_HAS_CXX14) && _HAS_CXX14 == 1)
-    #define JSON_HAS_CPP_14
 #endif
 
 /*!

--- a/single_include/nlohmann/json.hpp
+++ b/single_include/nlohmann/json.hpp
@@ -469,6 +469,14 @@ class other_error : public exception
     #endif
 #endif
 
+// C++ language standard detection
+#if (defined(__cplusplus) && __cplusplus >= 201703L) || (defined(_HAS_CXX17) && _HAS_CXX17 == 1) // fix for issue #464
+    #define JSON_HAS_CPP_17
+    #define JSON_HAS_CPP_14
+#elif (defined(__cplusplus) && __cplusplus >= 201402L) || (defined(_HAS_CXX14) && _HAS_CXX14 == 1)
+    #define JSON_HAS_CPP_14
+#endif
+
 // disable float-equal warnings on GCC/clang
 #if defined(__clang__) || defined(__GNUC__) || defined(__GNUG__)
     #pragma GCC diagnostic push
@@ -493,7 +501,11 @@ class other_error : public exception
 // allow for portable nodiscard warnings
 #if defined(__has_cpp_attribute)
     #if __has_cpp_attribute(nodiscard)
-        #define JSON_NODISCARD [[nodiscard]]
+        #if defined(__clang__) && !defined(JSON_HAS_CPP_17) // issue #1535
+            #define JSON_NODISCARD
+        #else
+            #define JSON_NODISCARD [[nodiscard]]
+        #endif
     #elif __has_cpp_attribute(gnu::warn_unused_result)
         #define JSON_NODISCARD [[gnu::warn_unused_result]]
     #else
@@ -544,14 +556,6 @@ class other_error : public exception
 #else
     #define JSON_LIKELY(x)      x
     #define JSON_UNLIKELY(x)    x
-#endif
-
-// C++ language standard detection
-#if (defined(__cplusplus) && __cplusplus >= 201703L) || (defined(_HAS_CXX17) && _HAS_CXX17 == 1) // fix for issue #464
-    #define JSON_HAS_CPP_17
-    #define JSON_HAS_CPP_14
-#elif (defined(__cplusplus) && __cplusplus >= 201402L) || (defined(_HAS_CXX14) && _HAS_CXX14 == 1)
-    #define JSON_HAS_CPP_14
 #endif
 
 /*!

